[Connectors API] Handle initialising connector default values in index service

jedrazb opened the pull request:
<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

### Changes
- Refactor `Connector` builder code, explicitly initialise default values in connector index service when creating a new connector. 
- Remove `null` checks in Connector class constructor. The `create connector` endpoint makes sure that those fields are non-null. Removing those null checks makes testing more flexible (e.g. SyncJob class references Connector class but doesn't care about e.g. `custom_scheduling`)
- Add more input validations to existing create connector endpoints, check for valid index name, add unit tests for validations

jedrazb committed:
[Connectors API] Handle initializing connector default values in index service

jedrazb committed:
Add index_name validations

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:Enterprise Search)

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -69,14 +71,23 @@ public ConnectorIndexService(Client client) {
      * @param listener  The action listener to invoke on response/failure.
      */
-    public void putConnector(String docId, Connector connector, ActionListener<DocWriteResponse> listener) {
+    public void putConnector(PutConnectorAction.Request request, ActionListener<DocWriteResponse> listener) {
```
Just a naming thing:

This can be addressed in a different PR, if you decide to address it 🙏  To me it's not obvious from a caller perspective, what the differences are between `putConnector` and `postConnector`. AFAIU both can be used to create or update a connector, but the sole difference lies in whether an id is auto-generated or not. I think the methods can be renamed to reflect that to the caller explicitly (you can also introduce a flag, but it's probably cleaner to have separate methods with clear names).

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/PostConnectorAction.java:
```diff
@@ -135,7 +139,18 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            ActionRequestValidationException validationException = null;
+
+            if (Strings.isNullOrEmpty(getIndexName())) {
+                validationException = addValidationError("index_name cannot be null or empty", validationException);
```
I saw at some other places, that field names and values are usually between brackets.
Example: https://github.com/elastic/elasticsearch/blob/e93892c0fdfd8e0614a67368d3f746fe033ccab1/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/FileOperatorUsersStore.java#L196-L206


```suggestion
                validationException = addValidationError("[index_name] cannot be [null] or [\"\"]", validationException);
```

Not too sure about `[null] or [\"\"]`, but you get the point 😄

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/PostConnectorAction.java:
```diff
@@ -135,7 +139,18 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                validationException = addValidationError("index_name cannot be null or empty", validationException);
+            }
+            try {
+                MetadataCreateIndexService.validateIndexOrAliasName(getIndexName(), InvalidIndexNameException::new);
```
Nice! ✍️  that there's something already built-in! 👏

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/PutConnectorAction.java:
```diff
@@ -156,6 +158,14 @@ public ActionRequestValidationException validate() {
                 validationException = addValidationError("connector_id cannot be null or empty", validationException);
             }
+            if (Strings.isNullOrEmpty(getIndexName())) {
+                validationException = addValidationError("index_name cannot be null or empty", validationException);
```
Same point as above, ofc nothing blocking 🙏

timgrein commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/action/PutConnectorActionTests.java:
```diff
@@ -0,0 +1,65 @@
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class PutConnectorActionTests extends ESTestCase {
```
Thanks for adding the tests 👏

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/UpdateConnectorLastSeenAction.java:
```diff
@@ -68,7 +63,7 @@ public ActionRequestValidationException validate() {
             builder.startObject();
             {
-                builder.field(Connector.LAST_SEEN_FIELD.getPreferredName(), lastSeen);
+                builder.field(Connector.ID_FIELD.getPreferredName(), connectorId);
```
Should we adapt the integration test using `is_after` similar to https://github.com/elastic/elasticsearch/blob/7f9ddd7f6ec517af9d4216f9a4c9b6b9038d0c9b/x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/420_connector_sync_job_check_in.yml#L17-L43 to make sure that the new timestamp is after the old one?

timgrein reviewed (commented):
Looking good! Added one question around an integration test. The rest is not blocking, happy to approve afterwards! 🚀

jedrazb committed:
Update x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/PostConnectorAction.java

Co-authored-by: Tim Grein <tim@4greins.de>

jedrazb committed:
Address review feedback

jedrazb commented:
Hey @timgrein thank you for your review! I addressed your feedback:
- used `is_after` assertion in yaml tests
- renamed put/post connector index service methods
- made error validation messages standard across all connector actions

jedrazb commented:
@elasticsearchmachine run elasticsearch-ci/bwc-snapshots

jedrazb committed:
Merge branch 'main' into connectors-api-refactor-connector-index-service

timgrein reviewed (approved):
Thanks for addressing the feedback! LGTM 🚢
